### PR TITLE
Migrate dataset code - episode II

### PR DESCRIPTION
**Pull request by banilo**
Back to the future of #647

### CODE REVIEW

**Comment by banilo** [comment]
Travis is happy, but coveralls complains (we did not so much change the amount of code, but mainly the code structure).

**Comment by AlexandreAbraham** [comment]
The only thing left is to rebase on master.

**Comment by banilo** [comment]
> The only thing left is to rebase on master.

I just did 5mins ago.

**Comment by AlexandreAbraham** [comment]
Sorry, I missed that! @GaelVaroquaux can we merge or do you have comments on the organization?

**Comment by banilo** [comment]
Gael said we can merge as soon as you, Alex, agree.

**Comment by banilo** [comment]
:hamster: Merge this very nice PR!
:imp: Hamster is right, there are now flaws left.
:rabbit: This is art!

**Comment by AlexandreAbraham** [comment]
Well, I'm afraid I'll have to wait for the approval of the octocat :-/

**Comment by AlexandreAbraham** [comment]
:octocat: not happy!
```
Traceback (most recent call last):
  File "/home/ubuntu/nilearn/doc/sphinxext/sphinxgallery/gen_rst.py", line 508, in generate_file_rst
    execfile(os.path.basename(src_file), my_globals)
  File "plot_nilearn_101.py", line 10, in <module>
    from nilearn import data
ImportError: cannot import name data
```
Oh noes! :octocat: devours :hamster: out of revenge because he was summoned in a broken PR!

**Comment by banilo** [comment]
> Oh noes

:octopus: sitting on :racehorse: ramping over :octocat: 

Ok, how do I fix this? By editing gen_rst.py with ```from nilearn.datasets import data```?

**Comment by banilo** [comment]
:v:

**Comment by AlexandreAbraham** [comment]
A special commission leaded by :shipit: is currently investigating your request.

**Comment by lesteve** [comment]
It'd be quite nice to investigate the coverage drop, 8 % is a big drop indeed !

**Comment by GaelVaroquaux** [comment]
Your tests are failing.

**Comment by lesteve** [comment]
Still failing I am afraid ...

File: nilearn/_utils/testing.py [hunk 1]
```diff
@@ -175,7 +175,8 @@ def mock_chunk_read_raise_error_(response, local_file, initial_size=0,
 
 
 class FetchFilesMock (object):
-    _mock_fetch_files = functools.partial(datasets._fetch_files, mock=True)
+    from nilearn.datasets.utils import _fetch_files
```
Comment (lesteve) [inline nilearn/_utils/testing.py]: Any reason why you put the import here and not at the top of the file?

Comment (lesteve) [inline nilearn/_utils/testing.py]: Also we want to always use relative imports for the nilearn ones.

File: setup.py [hunk 2]
```diff
@@ -84,9 +84,9 @@ def is_installing():
               'Programming Language :: Python :: 3.4',
           ],
           packages=find_packages(),
-          package_data={'nilearn.data': ['*.nii.gz', '*.csv'],
+          package_data={'nilearn.datasets.data': ['*.nii.gz', '*.csv'],
                         'nilearn.plotting.glass_brain_files': ['*.json'],
-                        'nilearn.tests.data': ['*'],
                         'nilearn.image.tests.data': ['*.mgz'],
-                        'nilearn.description': ['*.rst']},
+                        'nilearn.datasets.tests.data': ['*.*'],
```
Comment (lesteve) [inline setup.py]: Just put '*' rather than '*.*'

File: nilearn/_utils/testing.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/testing.py]: :+1:

File: nilearn/datasets/tests/test_atlas.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+    assert_true(isinstance(nibabel.load(ho.maps), nibabel.Nifti1Image))
+    assert_true(isinstance(ho.labels, np.ndarray))
+    assert_true(len(ho.labels) > 0)
+
+
+@with_setup(setup_mock, teardown_mock)
+@with_setup(setup_tmpdata, teardown_tmpdata)
+def test_fetch_atlas_craddock_2012():
+    print '*' * 80
```
Comment (AlexandreAbraham) [inline nilearn/datasets/tests/test_atlas.py]: This is not python 3 compatible + output are super annoying when running tests.

File: nilearn/datasets/tests/test_atlas.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    assert_raises_regex(ValueError, 'Invalid atlas name',
+                        atlas.fetch_atlas_harvard_oxford,
+                        'not_inside')
+
+    # specify existing atlas item
+    target_atlas = 'cort-maxprob-thr0-1mm'
+    target_atlas_fname = 'HarvardOxford-' + target_atlas + '.nii.gz'
+
+    HO_dir = os.path.join(tmpdir, 'fsl', 'data', 'atlases')
+    os.makedirs(HO_dir)
```
Comment (GaelVaroquaux) [inline nilearn/datasets/tests/test_atlas.py]: This should be named ho_dir: capitals are for classes.

File: nilearn/_utils/testing.py (hunk 1 above)
Comment (lesteve) [inline nilearn/_utils/testing.py]: Doing

```python
from ..datasets.utils import _fetch_files
```

at the top of your files seems to work fine for me.

Comment (lesteve) [inline nilearn/_utils/testing.py]: Also you now have an usused import on line 20.

**Comment by AlexandreAbraham** [comment]
Now you only need to address the last comments.

**Comment by GaelVaroquaux** [comment]
Merging! Thanks

**Pull request merged**